app/src/main/java/uk/gov/onelogin/login/ui/welcome/WelcomeScreenViewModel.kt

```diff
@@ -89,29 +96,38 @@ class WelcomeScreenViewModel @Inject constructor(
         } else {
             context.getString(R.string.openIdConnectClientId)
         }
-
         val scopes = listOf(LoginSessionConfiguration.Scope.OPENID)
-
         val locale = localeUtils.getLocaleAsSessionConfig()
-
         viewModelScope.launch {
             val persistentId = getPersistentId()?.takeIf { it.isNotEmpty() }
-            loginSession
-                .present(
-                    launcher,
-                    configuration = LoginSessionConfiguration(
-                        authorizeEndpoint = authorizeEndpoint,
-                        clientId = clientId,
-                        locale = locale,
-                        redirectUri = redirectUri,
-                        scopes = scopes,
-                        tokenEndpoint = tokenEndpoint,
-                        persistentSessionId = persistentId
-                    )
-                )
+            _loading.emit(true)
+            when (appIntegrity.startCheck()) {
+                is AppIntegrityResult.Failure -> handleAppIntegrityFailure()
+                else -> {
+                    _loading.emit(false)
+                    loginSession
+                        .present(
+                            launcher,
+                            configuration = LoginSessionConfiguration(
+                                authorizeEndpoint = authorizeEndpoint,
+                                clientId = clientId,
+                                locale = locale,
+                                redirectUri = redirectUri,
+                                scopes = scopes,
+                                tokenEndpoint = tokenEndpoint,
+                                persistentSessionId = persistentId
+                            )
+                        )
+                }
+            }
         }
     }
 
+    private suspend fun handleAppIntegrityFailure() {
+        _loading.emit(false)
```

Comment:
I have found it is nicer to have loading state changes limited to one method, as it is easier to not "miss" a state when making changes in the future. These particular methods are simple and it is visible as it can fit on one screen, but this becomes more important when dealing with a large number of state changes across a larger number of methods.